csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs

```diff
@@ -206,32 +209,108 @@ private void CanRunInferenceOnAModel(GraphOptimizationLevel graphOptimizationLev
                         validateRunResults(results);
                     }
                 }
+
+                // Run inference with pinned inputs and empty outputs
+                using (var pinnedInputs = new DisposableList<PinnedOnnxValue>())
+                {
+                    var inputNames = container.Select(i => i.Name).ToArray();
+                    pinnedInputs.AddRange(container.Select(i => PinnedOnnxValue.CreateFromTensor(i.AsTensor<float>())));
+
+                    // output names not specified
+                    using (var results = session.Run(inputNames, pinnedInputs))  // results is an IReadOnlyList<NamedOnnxValue> container
+                    {
+                        validateRunResults(results);
+                    }
+
+                    // output names specified explicitly
+                    using (var results = session.Run(inputNames, pinnedInputs, expectedOutputNames))  // results is an IReadOnlyList<NamedOnnxValue> container
+                    {
+                        validateRunResults(results);
+                    }
+                }
+
+                // Run inference with named inputs and named outputs
+                {
+                    // correct pre-allocated outputs
+                    var expectedOutputValues = new List<NamedOnnxValue>()
+                    {
+                        NamedOnnxValue.CreateFromTensor("softmaxout_1", new DenseTensor<float>(expectedOutputDimensions))
+                    };
+                    session.Run(container, expectedOutputValues);
+                    validateRunResultData(expectedOutputValues[0].AsTensor<float>());
+                }
+
+                // Run inference with pinned inputs and named outputs
+                using (var pinnedInputs = new DisposableList<PinnedOnnxValue>())
+                {
+                    var inputNames = container.Select(i => i.Name).ToArray();
+                    pinnedInputs.AddRange(container.Select(i => PinnedOnnxValue.CreateFromTensor(i.AsTensor<float>())));
```

Comment:
There are now  2 similar APIs: PinnedOnnxValue.CreateFromTensor() and NamedOnnxValue.CreateFromTensor().

How would a user know which to use (i.e. what are pros/cons of using PinnedOnnxValue)?

Reply:
`PinnedOnnxValue` should be used when literally user want to "pin" the value for reuse purpose. pros are all about lower latency for small models; cons include that user need to dispose the object manually after use, and it only support a determined-dimension if used as output.